# Predicting Customer Churn Rate in a Telecommunications Company

'Churn' is the number or percentage of customers who stop using a company's services during a given period. It is a key performance indicator, especially for subscription services, and is often used synonymously with "customer loss". A high churn rate can indicate problems with customer satisfaction, product quality, or pricing, and is a major concern because it directly impacts revenue and growth.

Project Overview: A telecommunications company wants to understand why customers are leaving its service and predict which customers are most likely to cancel their subscriptions next month.

## Data preparation

In [30]:
# load dataset
import pandas as pd
churn_data = pd.read_csv('./data/WA_Fn-UseC_-Telco-Customer-Churn.csv')
churn_data.sample(6)

customerID  gender  SeniorCitizen Partner Dependents  tenure  \
2419  2450-ZKEED  Female              0      No         No      11   
2520  5788-YPOEG  Female              0     Yes        Yes      34   
4624  0325-XBFAC    Male              0      No         No       8   
3080  1777-JYQPJ    Male              0      No         No       2   
429   6152-ONASV  Female              0     Yes         No      68   
6814  2267-WTPYD  Female              1     Yes         No      57   

     PhoneService     MultipleLines InternetService OnlineSecurity  \
2419          Yes                No             DSL             No   
2520          Yes                No             DSL            Yes   
4624          Yes                No     Fiber optic             No   
3080           No  No phone service             DSL             No   
429           Yes               Yes             DSL            Yes   
6814          Yes               Yes     Fiber optic             No   

     OnlineBackup DeviceProtection TechSupport StreamingTV StreamingMovies  \
2419           No              Yes         Yes          No              No   
2520          Yes              Yes         Yes         Yes             Yes   
4624           No              Yes          No         Yes             Yes   
3080           No               No          No          No              No   
429            No               No         Yes          No              No   
6814          Yes              Yes          No          No             Yes   

            Contract PaperlessBilling              PaymentMethod  \
2419        One year               No  Bank transfer (automatic)   
2520        Two year               No               Mailed check   
4624  Month-to-month              Yes           Electronic check   
3080  Month-to-month               No               Mailed check   
429         One year               No  Bank transfer (automatic)   
6814  Month-to-month               No    Credit card (automatic)   

      MonthlyCharges TotalCharges Churn  
2419           53.80       651.55    No  
2520           84.75      2839.45    No  
4624           94.70        740.3   Yes  
3080           24.30        38.45    No  
429            58.25       3975.7    No  
6814           94.00      5438.95    No

In [31]:
churn_data.shape

(7043, 21)

In [33]:
churn_data.columns

Index(['customerID', 'gender', 'SeniorCitizen', 'Partner', 'Dependents',
       'tenure', 'PhoneService', 'MultipleLines', 'InternetService',
       'OnlineSecurity', 'OnlineBackup', 'DeviceProtection', 'TechSupport',
       'StreamingTV', 'StreamingMovies', 'Contract', 'PaperlessBilling',
       'PaymentMethod', 'MonthlyCharges', 'TotalCharges', 'Churn'],
      dtype='object')

- customerID – Unique identifier for each customer.
- gender – Customer's gender (Male/Female).
- SeniorCitizen – Indicates whether the customer is a senior citizen (1 = Yes, 0 = No).
- Partner – Indicates if the customer has a partner (Yes/No).
- Dependents – Indicates if the customer has dependents (Yes/No).
- tenure – Number of months the customer has been with the company.
- PhoneService – Whether the customer subscribes to phone service (Yes/No).
- MultipleLines – Indicates if the customer has multiple phone lines (Yes/No/No phone service).
- InternetService – Type of internet service subscribed (DSL, Fiber optic, No).
- OnlineSecurity – Indicates whether the customer has online security add-on (Yes/No/No internet service).
- OnlineBackup – Indicates if the customer uses online backup service (Yes/No/No internet service).
- DeviceProtection – Indicates if the customer has device protection (Yes/No/No internet service).
- TechSupport – Indicates if the customer has tech support subscription (Yes/No/No internet service).
- StreamingTV – Indicates if the customer streams TV (Yes/No/No internet service).
- StreamingMovies – Indicates if the customer streams movies (Yes/No/No internet service).
- Contract – Type of contract (Month-to-month, One year, Two year).
- PaperlessBilling – Whether the customer uses paperless billing (Yes/No).
- PaymentMethod – Payment method used by the customer (Electronic check, Mailed check, Bank transfer, Credit card).
- MonthlyCharges – The amount charged to the customer monthly.
- TotalCharges – Total amount charged to the customer over the tenure.
- Churn – Indicates whether the customer has canceled the service (Yes/No).

In [34]:
churn_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   object 
 1   gender            7043 non-null   object 
 2   SeniorCitizen     7043 non-null   int64  
 3   Partner           7043 non-null   object 
 4   Dependents        7043 non-null   object 
 5   tenure            7043 non-null   int64  
 6   PhoneService      7043 non-null   object 
 7   MultipleLines     7043 non-null   object 
 8   InternetService   7043 non-null   object 
 9   OnlineSecurity    7043 non-null   object 
 10  OnlineBackup      7043 non-null   object 
 11  DeviceProtection  7043 non-null   object 
 12  TechSupport       7043 non-null   object 
 13  StreamingTV       7043 non-null   object 
 14  StreamingMovies   7043 non-null   object 
 15  Contract          7043 non-null   object 
 16  PaperlessBilling  7043 non-null   object 


If we look at the TotalCharges value, we can see that it was loaded incorrectly and loaded as an object instead of a float.

Let's fix it.

In [37]:
churn_data['TotalCharges'] = pd.to_numeric(churn_data['TotalCharges'], errors='coerce')

In [40]:
churn_data.describe()

SeniorCitizen       tenure  MonthlyCharges  TotalCharges
count    7043.000000  7043.000000     7043.000000   7032.000000
mean        0.162147    32.371149       64.761692   2283.300441
std         0.368612    24.559481       30.090047   2266.771362
min         0.000000     0.000000       18.250000     18.800000
25%         0.000000     9.000000       35.500000    401.450000
50%         0.000000    29.000000       70.350000   1397.475000
75%         0.000000    55.000000       89.850000   3794.737500
max         1.000000    72.000000      118.750000   8684.800000

In [41]:
churn_data.isna().sum()

customerID           0
gender               0
SeniorCitizen        0
Partner              0
Dependents           0
tenure               0
PhoneService         0
MultipleLines        0
InternetService      0
OnlineSecurity       0
OnlineBackup         0
DeviceProtection     0
TechSupport          0
StreamingTV          0
StreamingMovies      0
Contract             0
PaperlessBilling     0
PaymentMethod        0
MonthlyCharges       0
TotalCharges        11
Churn                0
dtype: int64

There are 11 NA values. Since these are few samples compared to the total number of samples, which is 7043, deleting these rows with the value NA in the TotalCharges column doesn't change anything; it's an insignificant sample.